Support ota-update tool (with cachix support)

avnik opened the pull request:
<!--
    Copyright 2022-2025 TII (SSRC) and the Ghaf contributors
    SPDX-License-Identifier: CC-BY-SA-4.0
-->

## Description of Changes

Host tool for update from cachix

### Type of Change
- [x] New Feature
- [ ] Bug Fix
- [ ] Improvement / Refactor

## Related Issues / Tickets

<!--
Link to GitHub issues or JIRA tickets (if any) that this PR addresses or is related to
-->
#1088 (remote execution stuff via admin was unstaged, currently only `ota-update` command is subject of testing)

## Checklist

<!--
Please check [X] for all items that apply. Leave [ ] if an item does not apply, but you have considered it.
Note that none of these are strict requirements — they are intended to inform reviewers.
Completing this checklist shows that you value and respect their time and effort.
-->

- [x] Clear summary in PR description
- [x] Detailed and meaningful commit message(s)
- [ ] Commits are logically organized and squashed if appropriate
- [x] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] Author has run `make-checks` and it passes
- [ ] All automatic GitHub Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [ ] Author has added reviewers and removed PR draft status

<!-- Additional description of omitted [ ] items if not obvious. -->

## Testing Instructions

### Applicable Targets
- [ ] Orin AGX `aarch64`
- [ ] Orin NX `aarch64`
- [x] Lenovo X1 `x86_64`
- [ ] Dell Latitude `x86_64`

### Installation Method
- [x] Requires full re-installation
- [x] Can be updated with `nixos-rebuild ... switch`
  (not sure about this)
- [ ] Other: 

### Test Steps To Verify:

#### Test 1
1. Install this release, ensure that network on host system is available
2. Ensure that `ota-update` command is available
3. Inspect bootloader state with `bootctl list`
4. Issue following command `ota-update cachix lenovo-x1-carbon-gen11-debug`. It should complete without errors
5. Inspect `bootctl list` again, it should contain new entry as candidate to boot
6. Reboot
7. Ensure that system booted into new config

#### Test 2
1. Install this release, ensure that network on host system is available
2. (on guivm) Invoke `givc-cli --name admin-vm --addr 192.168.100.3 --port 9001 --cacert /etc/givc/ca-cert.pem --cert /etc/givc/cert.pem --key /etc/givc/key.pem update list` to query installed updates
3. Invoke `givc-cli --name admin-vm --addr 192.168.100.3 --port 9001 --cacert /etc/givc/ca-cert.pem --cert /etc/givc/cert.pem --key /etc/givc/key.pem update cachix lenovo-x1-carbon-gen11-debug` to perform update
4. Invoke `givc-cli --name admin-vm --addr 192.168.100.3 --port 9001 --cacert /etc/givc/ca-cert.pem --cert /etc/givc/cert.pem --key /etc/givc/key.pem update list` again to query installed updates
5. Reboot
6. Ensure that system booted into new config
(test pins created for  gen11 only)

leivos-unikie commented:
Tested draft on Lenovo-X1, boot from USB SSD

`ota-update cachix lenovo-x1-carbon-gen11-debug` was finished with output
"could not acquire a lock"

After this

```
[root@ghaf-host:/var/empty]# bootctl list
         type: Boot Loader Specification Type #1 (.conf)
        title: NixOS (Generation 2 NixOS Xantusia 25.11.20250721.9fff9c5 (Linux 6.15.7), built on 2025-08-05) (default) (not reported/new)
           id: nixos-generation-2.conf
       source: /boot//loader/entries/nixos-generation-2.conf (on the EFI System Partition)
     sort-key: nixos
      version: Generation 2 NixOS Xantusia 25.11.20250721.9fff9c5 (Linux 6.15.7), built on 2025-08-05
   machine-id: fa0f1f863a6146f988e92b20d240af2d
        linux: /boot//EFI/nixos/rqryd1ncsdx6g3x6jq4xrvnfnbw4izi4-linux-6.15.7-bzImage.efi
       initrd: /boot//EFI/nixos/mpsngnz26gqixh6d6sg3nqvb56nqnksx-initrd-linux-6.15.7-initrd.efi
      options: init=/nix/store/730nzlw52hwsymyjxbk7qj2042ipprj2-nixos-system-ghaf-host-25.11.20250721.9fff9c5/init intel_iommu=on,sm_on iommu=pt module_blacklist=i915,xe,snd_pcm,bluetooth,bt>

         type: Boot Loader Specification Type #1 (.conf)
        title: NixOS (Generation 1 NixOS Xantusia 25.11.20250728.dc96378 (Linux 6.12.40), built on 2025-08-05) (selected)
           id: nixos-generation-1.conf
       source: /boot//loader/entries/nixos-generation-1.conf (on the EFI System Partition)
     sort-key: nixos
      version: Generation 1 NixOS Xantusia 25.11.20250728.dc96378 (Linux 6.12.40), built on 2025-08-05
   machine-id: fa0f1f863a6146f988e92b20d240af2d
        linux: /boot//EFI/nixos/ymgzw6vanvcq4cfycc8w0hxfgc4igq5p-linux-6.12.40-bzImage.efi
       initrd: /boot//EFI/nixos/niig0sysykgfbwl7nxxvyligmyc592gx-initrd-linux-6.12.40-initrd.efi
      options: init=/nix/store/y0v5j1n29kl0jyy5d4qy0dcd0jfhdm1a-nixos-system-ghaf-host-25.11.20250728.dc96378/init intel_iommu=on,sm_on iommu=pt module_blacklist=i915,xe,snd_pcm,bluetooth,bt>

         type: Automatic
        title: Reboot Into Firmware Interface
           id: auto-reboot-to-firmware-setup
       source: /sys/firmware/efi/efivars/LoaderEntries-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f (on the EFI System Partition)
```

After reboot NixOS menu listed 2 generations, booted to the new one

`ota-update` is not anymore available as expected

```
[root@ghaf-host:/var/empty]# bootctl list
         type: Boot Loader Specification Type #1 (.conf)
        title: NixOS (Generation 2 NixOS Xantusia 25.11.20250721.9fff9c5 (Linux 6.15.7), built on 2025-08-05) (default) (selected)
           id: nixos-generation-2.conf
       source: /boot//loader/entries/nixos-generation-2.conf (on the EFI System Partition)
     sort-key: nixos
      version: Generation 2 NixOS Xantusia 25.11.20250721.9fff9c5 (Linux 6.15.7), built on 2025-08-05
   machine-id: fa0f1f863a6146f988e92b20d240af2d
        linux: /boot//EFI/nixos/rqryd1ncsdx6g3x6jq4xrvnfnbw4izi4-linux-6.15.7-bzImage.efi
       initrd: /boot//EFI/nixos/mpsngnz26gqixh6d6sg3nqvb56nqnksx-initrd-linux-6.15.7-initrd.efi
      options: init=/nix/store/730nzlw52hwsymyjxbk7qj2042ipprj2-nixos-system-ghaf-host-25.11.20250721.9fff9c5/init intel_iommu=on,sm_on iommu=pt module_blacklist=i915,xe,snd_pcm,bluetooth,bt>

         type: Boot Loader Specification Type #1 (.conf)
        title: NixOS (Generation 1 NixOS Xantusia 25.11.20250728.dc96378 (Linux 6.12.40), built on 2025-08-05)
           id: nixos-generation-1.conf
       source: /boot//loader/entries/nixos-generation-1.conf (on the EFI System Partition)
     sort-key: nixos
      version: Generation 1 NixOS Xantusia 25.11.20250728.dc96378 (Linux 6.12.40), built on 2025-08-05
   machine-id: fa0f1f863a6146f988e92b20d240af2d
        linux: /boot//EFI/nixos/ymgzw6vanvcq4cfycc8w0hxfgc4igq5p-linux-6.12.40-bzImage.efi
       initrd: /boot//EFI/nixos/niig0sysykgfbwl7nxxvyligmyc592gx-initrd-linux-6.12.40-initrd.efi
      options: init=/nix/store/y0v5j1n29kl0jyy5d4qy0dcd0jfhdm1a-nixos-system-ghaf-host-25.11.20250728.dc96378/init intel_iommu=on,sm_on iommu=pt module_blacklist=i915,xe,snd_pcm,bluetooth,bt>

         type: Automatic
        title: Reboot Into Firmware Interface
           id: auto-reboot-to-firmware-setup
       source: /sys/firmware/efi/efivars/LoaderEntries-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f (on the EFI System Partition)
```

henrirosten commented on .github/pin-cachix-release.sh:
```diff
@@ -0,0 +1,27 @@
+PIN_NAME="${PREFIX}-%{TARGET}"
+
+path=$(nixos-rebuild build --flake ".#${TARGET}" --no-out-link)
+cachix push "${CACHIX_REPO}" "${path}"
+cachix pin "${CACHIX_REPO}" "${PIN_NAME}" "${path}"
```
Where does the cachix authentication token come from?

henrirosten commented on .github/workflows/pin-cachix-release.yml:
```diff
@@ -0,0 +1,44 @@
+
+name: eval
+on:
+  push:
+    branches:
+      - main
```
This is just for the WIP, right? I assume we don't want to pin all pushes to main?

avnik replied:
Why not? You suggest pin only stable tagged releases? Or may be use `ghaf-untrusted` repo for main

henrirosten replied:
There's only so much storage in our cachix cache. What happens when we pin more than we have storage?
We must have some process to delete the pins too, and remain somewhat selective as to what we choose to pin.

henrirosten reviewed (commented):
First quick comments below:

avnik commented:
@leivos-unikie address in test instruction was totally wrong, I updated test steps. Also removed statement about `ota-update` should disappear.  You could use `ota-update get` on host to ensure that system booted in proper version

leivos-unikie commented:
In Test 2 after installing fresh ghaf  I get

```
[ghaf@gui-vm:~]$ sudo givc-cli --name admin-vm --addr 192.168.100.3 --port 9001 --cacert /etc/givc/ca-cert.pem --cert /etc/givc/cert.pem --key /etc/givc/key.pem update list
2025-09-09T07:58:04.988973Z  INFO CLI is Cli {
    addr: "192.168.100.3",
    port: 9001,
    host_key: None,
    name: "admin-vm",
    vsock: None,
    cacert: Some(
        "/etc/givc/ca-cert.pem",
    ),
    cert: Some(
        "/etc/givc/cert.pem",
    ),
    key: Some(
        "/etc/givc/key.pem",
    ),
    notls: false,
    command: Update {
        update: List,
    },
}
2025-09-09T07:58:04.989269Z  INFO Connecting to https://192.168.100.3:9001, TLS name Some(TlsConfig { ca_cert_file_path: "/etc/givc/ca-cert.pem", cert_file_path: "/etc/givc/cert.pem", key_file_path: "/etc/givc/key.pem", tls_name: Some("admin-vm") })
2025-09-09T07:58:04.989465Z  INFO Using TLS name: admin-vm
Error: Detail: Exec error: Error: While read list of generations

Caused by:
    0: While decoding bootctl json output
    1: missing field `machineId`


Caused by:
    request contains invalid arguments
```

leivos-unikie commented:
It seems the list command was supposed to produce error before update. After update command it lists the update:

```
[ghaf@gui-vm:~]$ sudo givc-cli --name admin-vm --addr 192.168.100.3 --port 9001 --cacert /etc/givc/ca-cert.pem --cert /etc/givc/cert.pem --key /etc/givc/key.pem update cachix lenovo-x1-carbon-gen11-debug
2025-09-09T08:39:08.134883Z  INFO CLI is Cli {
    addr: "192.168.100.3",
    port: 9001,
    host_key: None,
    name: "admin-vm",
    vsock: None,
    cacert: Some(
        "/etc/givc/ca-cert.pem",
    ),
    cert: Some(
        "/etc/givc/cert.pem",
    ),
    key: Some(
        "/etc/givc/key.pem",
    ),
    notls: false,
    command: Update {
        update: Cachix(
            CachixOptions {
                pin_name: "lenovo-x1-carbon-gen11-debug",
                token: None,
                cache: "ghaf-dev",
                cachix_host: None,
            },
        ),
    },
}
2025-09-09T08:39:08.135111Z  INFO Connecting to https://192.168.100.3:9001, TLS name Some(TlsConfig { ca_cert_file_path: "/etc/givc/ca-cert.pem", cert_file_path: "/etc/givc/cert.pem", key_file_path: "/etc/givc/key.pem", tls_name: Some("admin-vm") })
2025-09-09T08:39:08.135288Z  INFO Using TLS name: admin-vm

[ghaf@gui-vm:~]$ sudo givc-cli --name admin-vm --addr 192.168.100.3 --port 9001 --cacert /etc/givc/ca-cert.pem --cert /etc/givc/cert.pem --key /etc/givc/key.pem update list
[sudo] password for ghaf: 
2025-09-09T08:45:20.253847Z  INFO CLI is Cli {
    addr: "192.168.100.3",
    port: 9001,
    host_key: None,
    name: "admin-vm",
    vsock: None,
    cacert: Some(
        "/etc/givc/ca-cert.pem",
    ),
    cert: Some(
        "/etc/givc/cert.pem",
    ),
    key: Some(
        "/etc/givc/key.pem",
    ),
    notls: false,
    command: Update {
        update: List,
    },
}
2025-09-09T08:45:20.254182Z  INFO Connecting to https://192.168.100.3:9001, TLS name Some(TlsConfig { ca_cert_file_path: "/etc/givc/ca-cert.pem", cert_file_path: "/etc/givc/cert.pem", key_file_path: "/etc/givc/key.pem", tls_name: Some("admin-vm") })
2025-09-09T08:45:20.254417Z  INFO Using TLS name: admin-vm
[Generation { generation: 2, date: "bogus", nixos_version: "25.11.20250903.5910558", kernel_version: "6.12.44", configuration_revision: "unknown", specialisations: [], current: false, store_path: "/nix/store/yah8s6rqsjh15yb5ihfmz6syb5wwvch5-nixos-system-ghaf-host-25.11.20250903.5910558" }, Generation { generation: 1, date: "bogus", nixos_version: "25.11.20250815.61cedad", kernel_version: "6.12.41", configuration_revision: "unknown", specialisations: [], current: true, store_path: "/nix/store/xp702aksgq8s0f8iich0fd2hf3b3falw-nixos-system-ghaf-host-25.11.20250815.61cedad" }]
```

leivos-unikie commented:
Both Test 1 and Test 2 passes now. 

They end up to this state:

```
[ghaf@ghaf-host:~]$ sudo bootctl list
         type: Boot Loader Specification Type #1 (.conf)
        title: NixOS (Generation 2 NixOS Xantusia 25.11.20250903.5910558 (Linux 6.12.44), built on 2025-09-09) (default) (selected)
           id: nixos-generation-2.conf
       source: /boot//loader/entries/nixos-generation-2.conf (on the EFI System Partition)
     sort-key: nixos
      version: Generation 2 NixOS Xantusia 25.11.20250903.5910558 (Linux 6.12.44), built on 2025-09-09
   machine-id: 3f9de86b94c94c8282aa67d778366c3c
        linux: /boot//EFI/nixos/jj5nrnb55b9pd5n0pkpywng0apvx2zmb-linux-6.12.44-bzImage.efi
       initrd: /boot//EFI/nixos/7ci469x2wr3qlgpd40f0hfklv48iwg52-initrd-linux-6.12.44-initrd.efi
      options: init=/nix/store/yah8s6rqsjh15yb5ihfmz6syb5wwvch5-nixos-system-ghaf-host-25.11.20250903.5910558/init quiet udev.log_priority=3 bgrt_disable=1 intel_iommu=on,sm_on iommu=pt module_blacklist=i915,xe,snd_pcm,bluetooth,>

         type: Boot Loader Specification Type #1 (.conf)
        title: NixOS (Generation 1 NixOS Xantusia 25.11.20250815.61cedad (Linux 6.12.41), built on 2025-09-04)
           id: nixos-generation-1.conf
       source: /boot//loader/entries/nixos-generation-1.conf (on the EFI System Partition)
     sort-key: nixos
      version: Generation 1 NixOS Xantusia 25.11.20250815.61cedad (Linux 6.12.41), built on 2025-09-04
   machine-id: 3f9de86b94c94c8282aa67d778366c3c
        linux: /boot//EFI/nixos/2z4rjg9v9fky6bcbs6dqb1d7v3d8n24m-linux-6.12.41-bzImage.efi
       initrd: /boot//EFI/nixos/271f2xic09v9fg9hdrv2j09ijxg723qz-initrd-linux-6.12.41-initrd.efi
      options: init=/nix/store/xp702aksgq8s0f8iich0fd2hf3b3falw-nixos-system-ghaf-host-25.11.20250815.61cedad/init intel_iommu=on,sm_on iommu=pt module_blacklist=i915,xe,snd_pcm,bluetooth,btusb acpi_backlight=vendor acpi_osi=linu>

         type: Automatic
        title: Reboot Into Firmware Interface
           id: auto-reboot-to-firmware-setup
       source: /sys/firmware/efi/efivars/LoaderEntries-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f (on the EFI System Partition)
```
